pkg/registry/core/service/storage/rest.go

```diff
@@ -824,41 +824,35 @@ func (al *RESTAllocStuff) initIPFamilyFields(oldService, service *api.Service) e
 		return nil // nothing more to do.
 	}
 
-	// Update-only prep work.
-	if oldService != nil {
-		np := service.Spec.IPFamilyPolicy
-
-		// If they didn't specify policy, or specified anything but
-		// single-stack AND they reduced these fields, it's an error.  They
-		// need to specify policy.
-		if np == nil || *np != api.IPFamilyPolicySingleStack {
-			el := make(field.ErrorList, 0)
-
-			if reducedClusterIPs(oldService, service) {
-				el = append(el, field.Invalid(field.NewPath("spec", "ipFamilyPolicy"), service.Spec.IPFamilyPolicy,
-					"must be 'SingleStack' to release the secondary cluster IP"))
-			}
-			if reducedIPFamilies(oldService, service) {
-				el = append(el, field.Invalid(field.NewPath("spec", "ipFamilyPolicy"), service.Spec.IPFamilyPolicy,
-					"must be 'SingleStack' to release the secondary IP family"))
-			}
-
-			if len(el) > 0 {
-				return errors.NewInvalid(api.Kind("Service"), service.Name, el)
-			}
-		} else { // policy must be SingleStack
-			// Update: As long as ClusterIPs and IPFamilies have not changed,
-			// setting policy to single-stack is clear intent.
-			if *(service.Spec.IPFamilyPolicy) == api.IPFamilyPolicySingleStack {
-				if sameClusterIPs(oldService, service) && len(service.Spec.ClusterIPs) > 1 {
-					service.Spec.ClusterIPs = service.Spec.ClusterIPs[0:1]
-				}
-				if sameIPFamilies(oldService, service) && len(service.Spec.IPFamilies) > 1 {
-					service.Spec.IPFamilies = service.Spec.IPFamilies[0:1]
-				}
-			}
+	// If the user didn't specify ipFamilyPolicy, we can infer a default.  We
+	// don't want a static default because we want to make sure that we never
+	// change between single- and dual-stack modes with explicit direction, as
+	// provided by ipFamilyPolicy.  Consider these cases:
+	//   * Create (POST): If they didn't specify a policy we can assume it's
+	//     always SingleStack.
+	//   * Update (PUT): If they didn't specify a policy we need to adopt the
+	//     policy from before.  This is better than always assuming SingleStack
+	//     because a PUT that changes clusterIPs from 2 to 1 value but doesn't
+	//     specify ipFamily would work.
+	//   * Update (PATCH): If they didn't specify a policy it will adopt the
+	//     policy from before.
+	if service.Spec.IPFamilyPolicy == nil {
+		if service.Spec.ClusterIP == api.ClusterIPNone && len(service.Spec.Selector) == 0 {
```

Comment:
```golang
// if we have old then it prevails against any defaulting (as per comments block earlier)
if oldService != nil && oldService.Spec.IPFamilyPolicy != nil {
// copy from old bail out
}

// there is no oldService || oldService.Spec.IPFamily ==nil
if service.Spec.ClusterIP == api.ClusterIPNone && len(service.Spec.Selector) == 0{
// set as require dual stack and bail out
}
// last case Create (with nil policy) or Update (with no prior policy set because it was externalName) 
// set as single stack and bail out.
```

Reply:
ACK